[Spring MVC] 방예혁 미션 제출합니다.

YehyeokBang opened the pull request:
## 설명
- `예외 처리` : 테스트 코드에서 확인하는 예외가 대부분 잘못된 값(인자)을 넘겼을 때 발생하는 예외라고 생각했습니다. 따라서, 별도로 커스텀 예외를 만들지 않고 `IllegalArgumentException` 표준 예외를 사용하여 코드의 간결성을 유지하고, 잘못된 인자로 인한 문제를 명확하게 표현하려고 했습니다.
- `불변 객체 사용` : `@RequestBody`를 사용하여 전달받는 객체는 클라이언트가 값을 입력하고 전달되는 과정에서 수정될 일이 없다고 판단하여 불변성을 가진 `Record`를 사용했습니다. 
- `ResponseEntity 사용` : 응답에 Location, Content-Type 등 헤더를 추가하기 위해 `ResponseEntity`를 사용했습니다.

YehyeokBang committed:
1단계-홈 화면 구현

- 웹, 타임리프 종속성을 추가합니다.
- home.html을 index.html로 이름을 변경하여 welcome page를 응답하도록 합니다.

YehyeokBang committed:
2단계-예약 조회 구현

- 예약 관리 페이지 조회 및 예약 목록 조회 API를 추가합니다.

YehyeokBang committed:
3단계-예약 추가 / 취소 구현

- 예약 추가 시 201 상태 코드 반환 및 Location 헤더를 추가합니다.
- 예약 취소 시 204 상태 코드를 반환합니다.

YehyeokBang committed:
4단계-예외 처리 구현

- 필요한 인자가 없거나 삭제할 예약이 없는 경우 400 상태 코드를 반환합니다.

hangillee commented on build.gradle:
```diff
@@ -14,6 +14,8 @@ repositories {
 
 dependencies {
     implementation 'org.springframework.boot:spring-boot-starter'
+    implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
+    implementation 'org.springframework.boot:spring-boot-starter-web'
```
알맞은 의존성 추가 👍

hangillee commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,62 @@
+@Controller
+public class ReservationController {
+    private List<Reservation> reservations = new ArrayList<>();
+    private AtomicLong index = new AtomicLong(1);
```
`AtomicLong` 사용 👍

hangillee commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,62 @@
+
+import static org.springframework.http.HttpStatus.CREATED;
+
+@Controller
+public class ReservationController {
+    private List<Reservation> reservations = new ArrayList<>();
+    private AtomicLong index = new AtomicLong(1);
+
+    @GetMapping("/reservation")
+    public String reservation() {
+        return "reservation";
+    }
+
+    @GetMapping("/reservations")
+    @ResponseBody
+    public ResponseEntity<List<Reservation>> reservations() {
+        return ResponseEntity.ok(reservations);
+    }
```
`@Controller`와 `@RestController`의 차이는 무엇인가요?
`@ResponseBody` 어노테이션은 무슨 역할을 가지고 있나요?

또한, View를 전달하는 메소드와 JSON을 반환하는 메소드를 한 곳에 모아둔 이유는 무엇인가요?

YehyeokBang replied:
`@Controller`는 Spring MVC에서 일반적인 웹 컨트롤러를 정의할 때 사용합니다! 주로 View 템플릿을 반환하는 메소드와 함께 사용되는데, `@Controller`를 사용하면 메소드에서 반환된 문자열은 뷰 리졸버(View Resolver)에 의해 해석되어 특정 Thymeleaf 등의 뷰 페이지로 포워딩됩니다. (예시: return "reservation" -> resources/templates/reservation.html)

`@RestController`는 RESTful 웹 서비스를 구축할 때 사용됩니다. `@RestController`는 `@Controller`와 `@ResponseBody`를 결합한 것과 동일한 역할을 합니다. 주로 JSON 또는 XML 형식의 데이터를 반환하는 API 엔드포인트를 정의할 때 사용되며, 메소드의 반환값은 자동으로 HTTP 응답의 바디에 직접 작성됩니다.

`@ResponseBody` 어노테이션이 메소드가 반환하는 값을 HTTP 응답의 바디에 직접 작성하는 역할을 수행합니다.

추가적으로 View를 전달하는 메서드와 JSON을 반환하는 메서드를 한 곳에 모아둔 이유는 현재 상태에서는 분리의 장점(더 작은 단위의 단일 책임 원칙, 추후 확장성 등)보다 통합의 장점(코드 관리, 직관적인 이해, 큰 틀로 봤을 때 예약이라는 도메인 묶음 등)이 더 크다고 생각했기 때문입니다!

hangillee commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,62 @@
+
+        Reservation newReservation = new Reservation(id, name, date, time);
+        reservations.add(newReservation);
+        return ResponseEntity.status(CREATED)
+                .header(HttpHeaders.LOCATION, "/reservations/" + id)
+                .header(HttpHeaders.CONTENT_TYPE, "application/json")
+                .body(newReservation);
```
적절한 상태 코드와 헤더 메타 데이터 👍

예혁이 201 CREATED를 사용한 이유는 무엇인가요?

YehyeokBang replied:
`201 CREATED` 상태코드와 함께 `Location` 헤더를 포함시켜서 새로운 리소스가 생성되었다는 것을 **명확히** 알리기 위해서 사용했습니다!

hangillee commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,62 @@
+
+    @PostMapping("/reservations")
+    @ResponseBody
+    public ResponseEntity<Reservation> createReservation(@RequestBody RequestReservation requestReservation) {
```
`@RequestBody`는 무슨 역할을 하나요?

YehyeokBang replied:
`@RequestBody`는 이 HTTP 요청 본문(body)을 자바 객체로 변환해주는 역할을 수행합니다.
예를 들어, JSON 형식의 데이터를 자바 클래스에 매핑해주기 때문에 개발자는 별도의 파싱 로직을 작성하지 않고도 요청 본문을 쉽게 객체로 다룰 수 있습니다.

개발자 입장에서는 `@ResponseBody`와 반대의 개념으로 이해할 수 있습니다! 두 어노테이션 모두 HTTP 요청과 응답에서 본문을 쉽게 컨트롤하기 위해 사용됩니다.

hangillee commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,62 @@
+
+    @DeleteMapping("/reservations/{reservationId}")
+    @ResponseBody
+    public ResponseEntity<Void> deleteReservations(@PathVariable Long reservationId) {
```
`@PathVariable`은 무슨 역할을 하나요?

YehyeokBang replied:
`@PathVariable` 어노테이션은 URL 경로의 일부를 변수로 사용할 수 있게 해줍니다. URL에 포함된 값을 메소드의 파라미터로 전달합니다. 여기서는 `{reservationId}` 부분에 해당하는 값을 Long 타입 reservationId로 전달하여 URL에 따라 동적으로 처리할 수 있게 도와줍니다.

특정 Id(요청 전까지는 모름)를 가진 예약을 처리할 수 있습니다.

hangillee commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,62 @@
+    public ResponseEntity<Void> deleteReservations(@PathVariable Long reservationId) {
+        boolean removed = reservations.removeIf(reservation -> reservation.getId().equals(reservationId));
+        if (!removed) {
+            throw new IllegalArgumentException("Reservation not found");
```
적절한 예외처리 👍
아직 사용자 정의(커스텀) 예외를 사용할 필요를 느끼지 못해 표준 예외를 사용했다는 설명은 아주 훌륭했습니다.

다만, 모든 상황에서 같은 예외를 던지면 모두 500 Internal Server Error로 보일 것입니다.
적절한 HTTP 상태 코드와 예외 내용 전달을 위해 예외 타입에 대해 고민해봅시다.

hangillee commented on src/main/java/roomescape/dto/RequestReservation.java:
```diff
@@ -0,0 +1,27 @@
+package roomescape.dto;
+
+public record RequestReservation(String name, String date, String time) {
+    private void checkName(String name) {
+        if (name == null || name.isBlank()) {
+            throw new IllegalArgumentException("Name is required");
+        }
+    }
+
+    private void checkDate(String date) {
+        if (date == null || date.isBlank()) {
+            throw new IllegalArgumentException("Date is required");
+        }
+    }
+
+    private void checkTime(String time) {
+        if (time == null || time.isBlank()) {
+            throw new IllegalArgumentException("Time is required");
+        }
+    }
```
적절한 Validation 👍

hangillee commented on src/main/java/roomescape/exception/ReservationControllerAdvice.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class ReservationControllerAdvice {
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity<String> IllegalArgumentException() {
+        return ResponseEntity.badRequest().build();
+    }
+}
```
전역 예외처리 👍
단순한 `ResponseEntity`를 반환하기보다 어디서 어떻게 예외가 발생했는지 추가 정보를 탑재해보는 건 어떨까요?

(참고) Advice라는 용어는 AOP에서 가져온 개념입니다!

hangillee commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -1,10 +1,16 @@
+
+import static org.hamcrest.CoreMatchers.is;
+
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
 @DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
```
`@SpringBootTest`와 `@DirtiesConext`를 처음 사용해보시나요?

각 어노테이션의 속성으로 지정된 `WebEnvironment.DEFINED_PORT`와 `ClassMode.BEFORE_EACH_TEST_METHOD`는 무엇을 의미하는 걸까요? 각 어노테이션에 대해 학습해보시고 저에게 알려주세요!

YehyeokBang replied:
`@SpringBootTest` 어노테이션은 Spring Boot 기반의 통합 테스트를 수행할 때 사용합니다. 이 어노테이션을 통해 스프링 애플리케이션 컨텍스트를 로드하고 테스트를 실행할 수 있습니다.

`SpringBootTest.WebEnvironment`는 테스트 시 웹 환경을 설정할 때 사용합니다.
- `MOCK` : Mock 서블릿 환경을 사용합니다. 실제 서블릿 컨테이너를 시작하지 않습니다.
- `RANDOM_PORT` : 임의의 포트를 사용하여 내장된 서버를 시작합니다.
- `DEFINED_PORT` : 애플리케이션 설정에 정의된 포트를 사용하여 내장된 서버를 시작합니다.
- `NONE` : 웹 환경을 시작하지 않습니다.
여기서 사용한 `WebEnvironment.DEFINED_PORT`는 애플리케이션 설정에 정의된 포트(일반적으로 application.properties 또는 application.yml 파일에 설정된 포트)를 사용하여 내장된 서버를 시작합니다. (실제로 스프링 부트를 실행시켜두고 테스트를 실행하면 포트 충돌이 발생합니다.) 이를 통해 실제 서버를 시작하여 테스트를 수행할 수 있습니다.

`@DirtiesContext` 어노테이션은 테스트 후에 해당 스프링 컨텍스트가 더러워졌음을 표시할 때 사용합니다. 이를 통해 스프링은 다음 테스트를 위해 새 컨텍스트를 로드합니다.

`DirtiesContext.ClassMode`는 테스트 클래스에서 컨텍스트를 정리하는 시점을 설정하는 데 사용됩니다.
- `BEFORE_CLASS` : 테스트 클래스의 모든 테스트 메서드 실행 전에 컨텍스트를 정리합니다.
- `AFTER_CLASS` : 테스트 클래스의 모든 테스트 메서드 실행 후에 컨텍스트를 정리합니다.
- `BEFORE_EACH_TEST_METHOD` : 각 테스트 메서드 실행 전에 컨텍스트를 정리합니다.
- `AFTER_EACH_TEST_METHOD` : 각 테스트 메서드 실행 후에 컨텍스트를 정리합니다.
사용한 `ClassMode.BEFORE_EACH_TEST_METHOD`는 각 테스트 메서드 실행 전에 컨텍스트를 정리하고, 새로운 컨텍스트를 로드합니다. 이를 통해 각 테스트가 독립적으로 실행되며, 서로 간섭하지 않도록 설정한 것입니다.

### 요약
`@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)` : 애플리케이션 설정에 정의된 포트를 사용하여 실제 서버를 시작하고 통합 테스트를 수행합니다.  
`@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)` : 각 테스트 메서드 실행 전에 스프링 컨텍스트를 정리하여 테스트 간의 간섭을 방지합니다.

hangillee replied:
훌륭하게 학습하셨네요! 👍

말씀해주신 것처럼 `@DirtiesContext`는 테스트 과정에서 지저분해진 Spring의 ApplicationContext를 완전히 초기화하여 재시작합니다. 현재 `@DirtiesContext`의 옵션은 `BEFORE_EACH_TEST_METHOD`입니다. 즉, 모든 테스트 메소드마다 **새로운** ApplicationContext가 시작된다는 이야기입니다.

이미 알고 계시겠지만 ApplicationContext를 구동하는데는 많은 리소스가 필요합니다.(스프링 어플리케이션 재시작하면 1~2초 정도의 딜레이가 있죠.) 만약, 테스트 메소드의 개수가 200개 가까이 된다면 무슨 문제가 발생할까요? `@DirtiesContext`가 과연 최선의 방법일까요? 🤔

YehyeokBang replied:
```java
// 스텝 테스트 코드 중 일부
RestAssured.given().log().all()
                .contentType(ContentType.JSON)
                .body(params)
                .when().post("/reservations")
                .then().log().all()
                .statusCode(201)
                .header("Location", "/reservations/1")
                .body("id", is(1));
```

미션에서 제공하는 테스트 코드 중 일부입니다. 보면 Location 헤더의 값이 `/reservations/1`인지 확인하는 부분이 있습니다. 만약 모든 스텝 테스트(1~4단계)를 실행할 때 컨텍스트를 공유한다면 달라지는 테스트 순서에 따라 실패할 수도 있는 테스트 코드인 것 같습니다. 
(예를 들어, 예약 생성 테스트가 여러 번 실행되면, 예약 ID가 예상과 다르게 증가하여 다른 테스트에서 의도하지 않은 결과를 초래할 수 있있을 것 같다는 생각입니다.)

## 1. 테스트 순서
그래서 `@TestMethodOrder(MethodOrderer.MethodName.class)`와 같은 어노테이션을 사용해서 순서를 직접 지정하는 방법도 생각해봤는데, 컨텍스트를 재사용하면서 빠르게 테스트를 성공시킨다는 관점에서는 좋아보이지만, 새로운 테스트 코드를 작성할 때 순서를 고려하면서 짜야한다는 부분이 더 많은 유지보수 대상을 만들 수 있다고 생각했습니다.

## 2. 테스트 코드를 수정
컨텍스트를 공유해도 문제가 없는 테스트를 작성하면 되지 않을까라는 생각을 했지만, 스텝 테스트는 미션 과정에서 제공하는 테스트이기 때문에 수정하지 않고, 추가로 다른 단위 테스트 형식으로 추가하면 좋을 것 같다고 생각했었습니다.

## 3. 질문에 대한 생각
> 이미 알고 계시겠지만 ApplicationContext를 구동하는데는 많은 리소스가 필요합니다.(스프링 어플리케이션 재시작하면 1~2초 정도의 딜레이가 있죠.) 만약, 테스트 메소드의 개수가 200개 가까이 된다면 무슨 문제가 발생할까요? @DirtiesContext가 과연 최선의 방법일까요?

공감가는 내용입니다. 하지만 2번과 같이 미션에서 제공하는 코드이기 때문에 원활한 미션 진행을 위해 그대로 두고, 새로운 단위 테스트를 작성하는 방향이 좋은 방법 같은데 리건은 어떻게 생각하시나요?

hangillee replied:
미션에서 제공하는 코드라고해서 수정하지 말라는 법은 없습니다. 😁 예혁이 생각하기에 더 나은 방법을 고민해보고 저를 설득해주시면 됩니다. 물론 `RestAssured`를 활용한 통합 테스트는 여러 테스트 방법의 학습을 위해서라도 유지하는게 좋을 것 같습니다.

그렇다면 어떻게 해야할까요? 예혁의 말대로 `@DirtiesContext` 없이는 각 테스트의 결과가 다른 테스트에 영향을 미칩니다. 현재 예혁은 가장 간편한 테스트 분리 방법인 `@DirtiesContext`의 문제점을 파악했습니다. 또한, 저는 테스트 메소드의 실행 순서를 강제로 지정하는 것도 좋지 못한 방법인 것 같습니다.(다른 개발자가 `@TestMethodOrder(MethodOrderer.MethodName.class)`를 잘못 이해할 가능성이 높습니다.) `@DirtiesContext`를 사용하지 않고 테스트를 분리할 수 있는 방법을 생각해봅시다!

hangillee commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -16,4 +22,74 @@ public class MissionStepTest {
                 .statusCode(200);
     }
+
+    @Test
+    void 이단계() {
```
적절한 DisplayName을 추가해서 해당 테스트가 무엇을 테스트하는지 명시해봅시다.
만약, 한 줄로 설명이 불가하다면 테스트를 적절히 분리해봅시다.

hangillee reviewed (changes requested):
안녕하세요, 예혁! 👋
즐거운 방학에 미션 수행하시느라 고생 많으셨습니다.

간단하게 리뷰 남겼으니 확인해주세요!

YehyeokBang committed:
예외 처리 보완

- 예외의 내용을 응답 본문에 추가합니다.

YehyeokBang committed:
테스트 메서드 이름 수정

- 메서드 이름은 영어로 작성하고 DisplayName 어노테이션을 사용해 가독성을 유지합니다.

YehyeokBang committed:
새로운 통합 테스트 추가

- 기존 스텝 테스트는 그대로 두고 새로운 통합 테스트 코드를 추가합니다.
- 각 테스트마다 새로운 컨텍스트를 사용하는 대신 하나의 컨텍스트로 테스트를 진행합니다.

YehyeokBang committed:
예약 생성 시 검증 로직 추가

- 생성자에 검증 로직을 추가하고 테스트 코드를 작성합니다.

YehyeokBang commented:
### 예외 관련
- 단순 BadRequest 응답이 아니라 본문에 예외 내용을 추가합니다.

### 테스트 관련
- 테스트 메서드 이름 영어로 수정, `@DisplayName` 사용으로 가독성 유지했습니다.
- `@DirtiesContext`를 사용하지 않는 통합 테스트 코드 추가
매 테스트마다 새로운 컨텍스트를 사용하는 스텝 테스트는 (원활한 미션 진행을 위해) 그대로 두고 하나의 컨텍스트를 공유하는 테스트 코드 작성했습니다.

### 기타
- dto 말고 예약 생성 시 자체 검증 로직 추가
도메인 스스로가 검증 로직을 갖고 있는 것이 맞다고 판단하여 생성자에 검증 로직을 추가했습니다. 하지만 아직까지는 검증 비즈니스가 다양하지 않기 때문에 dto와 도메인 모두 같은 검증 로직을 가지고 있습니다.
추가로 도메인 테스트 코드도 작성했습니다.

대부분 오프라인에서 의논했던 내용과 비슷합니다! @hangillee

hangillee commented on src/main/java/roomescape/model/Reservation.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    public String getDate() {
+        return date;
+    }
+
+    public String getTime() {
+        return time;
+    }
+
+    private void checkAllFields(Long id, String name, String date, String time) {
+        checkId(id);
+        checkName(name);
+        checkDate(date);
+        checkTime(time);
+    }
+
+    private void checkId(Long id) {
+        if (id == null) {
+            throw new IllegalArgumentException("Id is required");
+        }
+    }
+
+    private void checkName(String name) {
+        if (name == null || name.isBlank()) {
+            throw new IllegalArgumentException("Name is required");
+        }
+    }
+
+    private void checkDate(String date) {
+        if (date == null || date.isBlank()) {
+            throw new IllegalArgumentException("Date is required");
+        }
+    }
+
+    private void checkTime(String time) {
+        if (time == null || time.isBlank()) {
+            throw new IllegalArgumentException("Time is required");
+        }
+    }
```
사소하지만, `Getter`보단 엔티티 생성 검증 로직들이 위에 오는 것이 좋습니다.

`Getter`는 누구나 명확하게 구현 내용을 알 수 있지만 검증 로직은 읽어야 하거든요.

hangillee commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -1,19 +1,100 @@
     }
+
+    @Test
+    @DisplayName("2단계-예약 목록 조회 테스트")
```
명확한 테스트 목적 명시 👍

hangillee commented on src/test/java/roomescape/model/ReservationTest.java:
```diff
@@ -0,0 +1,29 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class ReservationTest {
+    @Test
+    @DisplayName("필드가 하나라도 null이면 IllegalArgumentException 발생")
+    void create() {
+        assertThrows(IllegalArgumentException.class, () -> new Reservation(null, "bang", "2024-07-01", "10:00"));
+        assertThrows(IllegalArgumentException.class, () -> new Reservation(1L, null, "2024-07-01", "10:00"));
+        assertThrows(IllegalArgumentException.class, () -> new Reservation(1L, "bang", null, "10:00"));
+        assertThrows(IllegalArgumentException.class, () -> new Reservation(1L, "bang", "2024-07-01", null));
+    }
+
+    @Test
+    @DisplayName("모든 필드가 정상적으로 입력되면 Reservation 객체 생성")
+    void create1() {
+        Reservation reservation = assertDoesNotThrow(
+                () -> new Reservation(1L, "bang", "2024-07-01", "10:00")
+        );
+        assertEquals(1L, reservation.getId());
+        assertEquals("bang", reservation.getName());
+        assertEquals("2024-07-01", reservation.getDate());
+        assertEquals("10:00", reservation.getTime());
+    }
+}
```
도메인 단위 테스트 👍

hangillee commented on src/test/java/roomescape/RoomEscapeTest.java:
```diff
@@ -0,0 +1,129 @@
+                .header("Location", matchesPattern("/reservations/\\d+"))
+                .extract().response();
+
+        // Location 헤더에서 ID 추출
+        String locationHeader = response.getHeader("Location");
+        String extractedId = locationHeader.replaceAll(".*/reservations/", "");
+
+        // Location 헤더에 있는 ID와 응답 본문의 ID가 일치하는지 확인
+        response.then()
+                .body("id", is(Integer.parseInt(extractedId)))
+                .body("name", is("bang"))
+                .body("date", is("2024-07-01"))
+                .body("time", is("15:00"));
```
응답 URI의 값 통해 예약 추가 API를 검증하는 방식은 꽤 새로운 것 같습니다! 멋지네요.

여담이지만, 나중에 DB를 추가하게 됐을 때는 **DB의 테이블도 초기화**하는 로직이 필요해집니다! 😁
추후 미션에서 `@DirtiesContext`를 사용하지 않고 테스트를 분리할 수 있는 방법에 대해 고민해봅시다.

hangillee reviewed (approved):
안녕하세요, 예혁! 🙌
리뷰 잘 반영해주셨네요.

제가 드린 질문에 대한 답변도 깊게 고민해보신 흔적이 보여 제가 다 뿌듯합니다.
앞으로 남은 미션들도 파이팅입니다! 이만 머지 요청하겠습니다!
